rust: make pkgsCross.wasi32.rustPlatform.buildRustPackage work

jcaesar opened the pull request:
## Description of changes

This PR makes it possible to e.g. build and run rust packages like `nix shell .#pkgsCross.wasi32.charasay -c charasay.wasm say moo` (assuming you have `wasm32-wasi` in your emulated systems).

Without this PR, building the cross rustc fails with
```
thread 'main' panicked at src/core/build_steps/compile.rs:370:17:
Target "wasm32-unknown-wasi" does not have a "wasi-root" ```
```

I think this is neat, but it admittedly isn't super useful, as a lot of packages fail to compile due to failing C dependencies or some common dependencies assuming that all wasm is for the web. I've tested the 469 rust packages in by-name, and only 60¹ compile. Even those that do compile might fail at runtime, e.g. due to std::thread not being supported and panicking when attempting to create a thread (like alejandra).

I'm only aware of two mildly useful things that work, namely rq and qrtool. But I think it's nice to have this working.

¹ aarch64-esr-decoder, aba, action-validator, adrs, bunbun, cargo-autoinherit, cargo-bloat, cargo-bump, cargo-shear, cargo-swift, catppuccin-whiskers, catppuccinifier-cli, cdwe, circom, clippy-sarif, cmd-wrapped, commitmsgfmt, diesel-cli-ext, dipc, elf-info, emacs-lsp-booster, esbuild-config, french-numbers, galerio, git-upstream, hadolint-sarif, hvm, hyperlink, icnsify, ifrextractor-rs, keedump, kickstart, ldproxy, little_boxes, ludtwig, lutgen, majima, mcumgr-client, mdsh, ndstrim, obsidian-export, okolors, parallel-disk-usage, pdfrip, polylux2pdfpc, preserves-tools, prettypst, protoc-gen-rust-grpc, qrtool, rq, rs-tftpd, rusti-cal, rusty-diceware, sarif-fmt, shellcheck-sarif, tera-cli, thud, toml-cli, typstfmt, unison-fsmonitor, wit-bindgen

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jcaesar committed:
rustc: set up sysroot for wasm32-wasi and fix build

jcaesar commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -145,14 +146,19 @@ in stdenv.mkDerivation (finalAttrs: {
   ] ++ optionals stdenv.targetPlatform.isMusl [
     "${setTarget}.musl-root=${pkgsBuildTarget.targetPackages.stdenv.cc.libc}"
+  ] ++ optionals stdenv.targetPlatform.isWasi [
+    "${setTarget}.wasi-root=${runCommandLocal "wasi-sysroot" {} ''
+      mkdir -p $out/lib
+      ln -s ${pkgsBuildTarget.targetPackages.stdenv.cc.libc}/lib $out/lib/wasm32-wasi
+    ''}"
```
This workaround is required because wasilibc [overrides](https://github.com/NixOS/nixpkgs/blob/8fa00a61925cf7fe2c302b8d0d9bb5119a7388bb/pkgs/development/libraries/wasilibc/default.nix#L33) the default path/folder structure it's being installed to, but rust expects that structure to exist. The rustc build copies the files from that folder, so this derivation won't get referenced in the closure.

Alternatively this could be done by
1. creating a symlink with the right folder stucture in wasilibc's postInstall (which isn't pretty either and would cause a rebuild of firefox)
2. creating some tmp directory with the right folder structure during rustc's postPatch (but that would mean this workaround must be spread over two different places in this file, which I'd prefer to avoid.

So I think this is the prettiest way of handling that.

alyssais replied:
> creating a symlink with the right folder stucture in wasilibc's postInstall like [so](https://github.com/NixOS/nixpkgs/compare/8aa5aa70d620106e0be3cab003d5cd0c49fae874..d57e347cee6e74743e019b0b30bb2a3230837b96) (which isn't pretty either and would mean I have to touch something that's in the build path of firefox, which I'd like to avoid)

I think this makes more sense.  If Rust expects it here, other things might too.  This is not an urgent change, so there's definitely no reason avoiding a one-off Firefox rebuild should be a consideration.

jcaesar replied:
If you say "other things too", I should probably create that symlink for the headers, too. Done.

jcaesar commented on pkgs/build-support/rust/lib/default.nix:
```diff
@@ -23,7 +23,9 @@ rec {
     cxxForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++";
-    linkerForHost = if shouldUseLLD stdenv.targetPlatform
+    linkerForHost = if stdenv.targetPlatform.isWasi
+      then "${pkgsBuildHost.llvmPackages.bintools-unwrapped}/bin/${stdenv.cc.targetPrefix}wasm-ld"
```
The problem here is that
1. wasm-ld isn't included by bintools-wrapper. And even if add that, rustc seems to pass `-flavor wasm` to the linker, even though its path ends in `wasm-ld`.)
2. `ld.lld` from `bintools-unwrapped` would also work, but from what I understand, it works because rustc passes `-flavor wasm`, which is a legacy option. `ld.lld` from `bintools` chokes on that argument.
3. Using  `"${stdenv.cc}/bin/${stdenv.cc.targetPrefix}clang"` as a linker could work (must be "clang", not "cc" as used in `ccForHost`), but I have been unable to figure out how to explain to clang that the linker it should use is not named `wasm-ld` but `wasm32-unknown-wasi-wasm-ld` (it does have that in its search path already).

alyssais replied:
Wouldn't using a wrapped linker be necessary to find dependencies?  What linker does Rust use by default?

jcaesar replied:
> What linker does Rust use by default?

Either `"${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc"` if that's clang or ld.lld otherwise, both are wrapped, from what I can see.
> Wouldn't using a wrapped linker be necessary to find dependencies?

It's possible that there isn't a single rust package that depends on nixpkgs-built dependencies makes it to linking. Most of them do depend on stuff that doesn't work on wasi, like openssl or gui-related libraries. So I can't answer that question. Why using a wrapped linker breaks certainly needs some investigation… Not sure I have time for that.

Mic92 replied:
What are the observed symptoms of it not working? Not that if you se NIX_DEBUG than it will also what flags it injects into the build.

Mic92 replied:
I would at least leave a comment why we are doing this for now, potentially combined with the error message.

jcaesar replied:
@Mic92 Thank you for the `NIX_DEBUG` hint. Getting rust to print the linker command line in the unwrapped case was a bit more tricky, but with both, I was able to confirm what's going on.

The wrapper script was passing hardening flags (`-z relro -z now`) to wasm-ld, which 

* cause lld to not recognize the `-flavor wasm` arguments
* don't make sense on wasm (`warning: unknown -z value: now`) anyway

I'm not sure what's best here, I don't think hardening makes a lot of sense when compiling to wasm anyway, but I don't know of a way to disable hardening for an entire target. What I opted for is to instead add `hardening_unsupported_flags+=" relro bindnow"` in the wrapper script on wasi.

---
For reference:

The output the linker wrapper is:
```
error: linking with `/nix/store/6plp6qym01fbp05lwa1rh4ym50vnw1ng-wasm32-unknown-wasi-llvm-binutils-wrapper-17.0.6/bin/wasm32-unknown-wasi-wasm-ld` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/x86_64-unknown-linux-gnu/bin:/nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/nix/store/708hwm
  = note: HARDENING: disabled flags: pie
          HARDENING: Is active (not completely disabled with "all" flag)
          HARDENING: enabling relro
          HARDENING: enabling bindnow
          extra flags before to /nix/store/5gghgw5dkhw3dx862wynrd26sm3k8xbp-llvm-binutils-17.0.6/bin/wasm32-unknown-wasi-wasm-ld:
            -z
            relro
            -z
            now
          original flags to /nix/store/5gghgw5dkhw3dx862wynrd26sm3k8xbp-llvm-binutils-17.0.6/bin/wasm32-unknown-wasi-wasm-ld:
            -flavor
            wasm
            --rsp-quoting=posix
            --export
            __main_void
            -z
            stack-size=1048576
            --stack-first
            --allow-undefined
            --no-demangle
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib/self-contained/crt1-command.o
            /build/source/target/wasm32-wasip1/release/deps/rq-9dc42ad07b27356d.rq.2f192189ed26cc41-cgu.0.rcgu.o
            -L
            /build/source/target/wasm32-wasip1/release/deps
            -L
            /build/source/target/release/deps
            -L
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib
            -l
            c
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib/libcompiler_builtins-dae8cdcfb0b90f99.rlib
            -L
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib
            -L
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib/self-contained
            -o
            /build/source/target/wasm32-wasip1/release/deps/rq-9dc42ad07b27356d.wasm
            --gc-sections
            -O3
          extra flags after to /nix/store/5gghgw5dkhw3dx862wynrd26sm3k8xbp-llvm-binutils-17.0.6/bin/wasm32-unknown-wasi-wasm-ld:
            -L/nix/store/56c3y4h4ba85fxz0876ld9qd8l3ww2hy-libcxx-static-wasm32-unknown-wasi-17.0.6/lib
            -L/nix/store/02cizbm0w4akb46mapwb0in8jcr10srp-wasilibc-static-wasm32-unknown-wasi-21/lib
          wasm32-unknown-wasi-wasm-ld: error: unknown argument: -flavor
          wasm32-unknown-wasi-wasm-ld: warning: unknown -z value: relro
          wasm32-unknown-wasi-wasm-ld: warning: unknown -z value: now
          wasm32-unknown-wasi-wasm-ld: error: cannot open wasm: No such file or directory
```

This is what I was trying with:

```diff
diff --git a/pkgs/build-support/bintools-wrapper/default.nix b/pkgs/build-support/bintools-wrapper/default.nix
index e7fcf173c602..1c955fdd8d0d 100644
--- a/pkgs/build-support/bintools-wrapper/default.nix
+++ b/pkgs/build-support/bintools-wrapper/default.nix
@@ -239,6 +239,8 @@ stdenvNoCC.mkDerivation {
         basename=$(basename "$variant")
         wrap $basename ${./ld-wrapper.sh} $variant
       done
+    '' + optionalString targetPlatform.isWasi ''
+      wrap ${targetPrefix}wasm-ld ${./ld-wrapper.sh} $ldPath/${targetPrefix}wasm-ld
     '';
 
   strictDeps = true;
diff --git a/pkgs/build-support/rust/lib/default.nix b/pkgs/build-support/rust/lib/default.nix
index b52eecccbb2f..7f66bf01c1e5 100644
--- a/pkgs/build-support/rust/lib/default.nix
+++ b/pkgs/build-support/rust/lib/default.nix
@@ -24,7 +24,7 @@ rec {
     ccForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc";
     cxxForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++";
     linkerForHost = if stdenv.targetPlatform.isWasi
-      then "${pkgsBuildHost.llvmPackages.bintools-unwrapped}/bin/${stdenv.cc.targetPrefix}wasm-ld"
+      then "${pkgsBuildHost.llvmPackages.bintools}/bin/${stdenv.cc.targetPrefix}wasm-ld"
       else if shouldUseLLD stdenv.targetPlatform
       && !stdenv.cc.bintools.isLLVM
       then "${pkgsBuildHost.llvmPackages.bintools}/bin/${stdenv.cc.targetPrefix}ld.lld"
diff --git a/pkgs/development/tools/rq/default.nix b/pkgs/development/tools/rq/default.nix
index f8ed521ad2cc..589d347dbac5 100644
--- a/pkgs/development/tools/rq/default.nix
+++ b/pkgs/development/tools/rq/default.nix
@@ -24,6 +24,9 @@ rustPlatform.buildRustPackage rec {
     rm build.rs
   '';
 
+  env.NIX_DEBUG = "1";
+  env.RUSTC_LOG = "rustc_codegen_ssa::back::link=info";
+
   VERGEN_SEMVER = version;
 
   meta = with lib; {
```

alyssais replied:
Hardening flags are already platform-dependent, so I think you'd just need to add a case for wasmi.

alyssais replied:
By "by default", I mean what linker does a rustc you'd get from rustup use in this case?

jcaesar replied:
Ah, I see. @alyssais For the docker.io/library/rust docker image (which is currently on 1.79) it defaults to `rust-lld -flavor wasm`.

jcaesar replied:
(I'm sorry, I force-pushed over this commit again, I mistakenly thought not changing the linker was going to work. But it does have to be set to lld.)

jcaesar commented:
I've re-verified that this is the best solution I can come up with. I've added comments on the  @Kranzes, if you only removed your reviewer status because this was still a draft, can I ask you to have a look at some point?

alyssais commented on lib/systems/examples.nix:
```diff
@@ -353,6 +353,7 @@ rec {
 
   wasi32 = {
     config = "wasm32-unknown-wasi";
+    rustc.config = "wasm32-wasi";
```
The Rust documentation says "undergoing a [rename to wasm32-wasip1](https://github.com/rust-lang/compiler-team/issues/607)".  Should we use the new name?  Also, should we use the threads variant?

jcaesar replied:
That's only available from Rust 1.78, nixpkgs is still on 1.77. Btw, wasm32-wasi will go away January 2025. [cf this timeline](https://blog.rust-lang.org/2024/04/09/updates-to-rusts-wasi-targets.html#timeline)

alyssais replied:
… and 1.78 on master

jcaesar replied:
Must have missed that it hit master... Anyway, rebased on master and updated. Thank you.

Problem is that now

jcaesar commented:
Thank you for the review.

(The reason I wasn't particularly keen on touching wasilibc is that the following takes ages to build. The floorp failure also happens on master, so that's unrelated.)

---

Result of `nixpkgs-review pr 323161` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>floorp</li>
    <li>floorp-unwrapped</li>
    <li>floorp-unwrapped.debug</li>
  </ul>
</details>
<details>
  <summary>32 packages built:</summary>
  <ul>
    <li>adoptopenjdk-icedtea-web</li>
    <li>betterbird</li>
    <li>betterbird-unwrapped</li>
    <li>betterbird-unwrapped.debug</li>
    <li>eyewitness</li>
    <li>firefox</li>
    <li>firefox-beta</li>
    <li>firefox-beta-unwrapped</li>
    <li>firefox-beta-unwrapped.debug</li>
    <li>firefox-beta-unwrapped.symbols</li>
    <li>firefox-devedition</li>
    <li>firefox-devedition-unwrapped</li>
    <li>firefox-devedition-unwrapped.debug</li>
    <li>firefox-devedition-unwrapped.symbols</li>
    <li>firefox-esr</li>
    <li>firefox-esr-115-unwrapped</li>
    <li>firefox-esr-115-unwrapped.debug</li>
    <li>firefox-esr-115-unwrapped.symbols</li>
    <li>firefox-mobile</li>
    <li>firefox-unwrapped</li>
    <li>firefox-unwrapped.debug</li>
    <li>firefox-unwrapped.symbols</li>
    <li>firefoxpwa</li>
    <li>librewolf</li>
    <li>librewolf-unwrapped</li>
    <li>librewolf-unwrapped.debug</li>
    <li>sitespeed-io</li>
    <li>slimerjs</li>
    <li>thunderbird</li>
    <li>thunderbird-unwrapped</li>
    <li>thunderbird-unwrapped.debug</li>
    <li>thunderbird-unwrapped.symbols</li>
  </ul>
</details>

alyssais commented on pkgs/build-support/rust/build-rust-package/default.nix:
```diff
@@ -45,7 +45,7 @@
 , useNextest ? false
 # Enable except on aarch64 pkgsStatic, where we use lld for reasons
-, auditable ? !cargo-auditable.meta.broken && !(stdenv.hostPlatform.isStatic && stdenv.hostPlatform.isAarch64 && !stdenv.hostPlatform.isDarwin)
+, auditable ? !cargo-auditable.meta.broken && (!stdenv.hostPlatform.isStatic || !stdenv.hostPlatform.isAarch64 || stdenv.hostPlatform.isDarwin) && !stdenv.hostPlatform.isWasi
```
I think the motivation for the logic here is more difficult to understand with the ! distributed inside.  Why change that?

Would probably also be clearer to move the wasi case before the complicated one with the parentheses, and maybe start a new line after each top-level `&&`.

jcaesar replied:
I wasn't able to wrap my head around what I wanted to do without putting this into negation normal form (personal preference? double negation is mean). The motivation is to disable auditable in "this *or* that" situation, yet the top level operator is an &&.

I considered doing this

```diff
diff --git a/pkgs/build-support/rust/build-rust-package/default.nix b/pkgs/build-support/rust/build-rust-package/default.nix
index 36e0408cc198..d85820ab536a 100644
--- a/pkgs/build-support/rust/build-rust-package/default.nix
+++ b/pkgs/build-support/rust/build-rust-package/default.nix
@@ -44,9 +44,14 @@
 , buildFeatures ? [ ]
 , checkFeatures ? buildFeatures
 , useNextest ? false
-# Enable except on aarch64 pkgsStatic, where we use lld for reasons
-, auditable ? !cargo-auditable.meta.broken && !(stdenv.hostPlatform.isStatic && stdenv.hostPlatform.isAarch64 && !stdenv.hostPlatform.isDarwin)
-
+, auditable ? !( # rust binaries are made auditable except for any of these:
+    # - cargo-auditable is broken
+    cargo-auditable.meta.broken
+    # - we're on wasi
+    || stdenv.hostPlatform.isWasi
+    # - were on static aarch64 except darwin
+    || (stdenv.hostPlatform.isStatic && stdenv.hostPlatform.isAarch64 && !stdenv.hostPlatform.isDarwin)
+  )
 , depsExtraArgs ? {}

 # Toggles whether a custom sysroot is created when the target is a .json file.
 ```

but not doing this at all just causes a little warning at the end of the build

```
WARNING: target 'wasm32-wasip1' is not supported by 'cargo auditable'!
The build will continue, but no audit data will be injected into the binary.
```

Ignoring that warning and not touching this line at all seems like the lesser evil.

alyssais replied:
Agreed.

alyssais commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -19,6 +19,7 @@
 , fastCross
 , lndir
 , makeWrapper
+, runCommandLocal
```
No longer used.

jcaesar replied:
Thank you.

alyssais commented:
I tried building pkgsCross.wasi32.qrtool, but it failed to build wasm32-unknown-wasi-rustc-1.78.0.

```
thread 'main' panicked at src/core/build_steps/compile.rs:400:17:
Target "wasm32-unknown-wasi" does not have a "wasi-root" key in Config.toml
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:00:13
```

jcaesar commented:
@alyssais Could you tell me exactly what you're trying to build? The following work for me (should be the same thing):

```
nix build  -vL github:NixOS/nixpkgs/56df69358a7c06f415c33c3cb9543e41b872c220#pkgsCross.wasi32.rq
nix build  -vL github:NixOS/nixpkgs/pull/323161/head#pkgsCross.wasi32.rq
```

(qrtool will fail in installPhase because of #289517 / #308283. Try rq instead if the above does get past the failure you saw in rustc.)

alyssais commented:
Looks like I might have mixed up which branch I was building — sorry!

alyssais commented on pkgs/build-support/rust/lib/default.nix:
```diff
@@ -23,8 +23,9 @@ rec {
     cxxForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++";
-    linkerForHost = if shouldUseLLD stdenv.targetPlatform
-      && !stdenv.cc.bintools.isLLVM
+    # For wasi: rustc passes lld-flavored args unless the compiler is named "clang". So use lld.
```
Do you know which code in rustc does this check?

jcaesar replied:
After some digging. I think I found the [relevant part](https://github.com/rust-lang/rust/blob/adeb79d3f523dfc11a8e9db64ddfc5f546f650bb/compiler/rustc_target/src/spec/mod.rs#L314-L324). So `clang` or `gcc` as names would be recognized, but not `cc` (as set in `ccForHost`). I've revised the comment accordingly.
(Defaulting to lld without being wrapped in a compiler is [here](https://github.com/rust-lang/rust/blob/adeb79d3f523dfc11a8e9db64ddfc5f546f650bb/compiler/rustc_target/src/spec/base/wasm.rs#L90).)

Side notes:
* Since `cc` isn't recognized as anything by rust, the current setup for `ccForHost` is maybe not the best in general.
* [This doccoment](https://doc.rust-lang.org/beta/nightly-rustc/rustc_target/spec/enum.LinkerFlavor.html) has some helpful insights.

alyssais replied:
> * Since `cc` isn't recognized as anything by rust, the current setup for `ccForHost` is maybe not the best in general…?

Yeah, that's what I was wondering.

alyssais replied:
It might also make sense to teach Rust to understand "cc", as it's not exactly a Nixpkgs-specific convention.

jcaesar replied:
I think I read somewhere that not recognizing cc is intentional. Somebody felt that two letters is too little to go on? I might be mistaken, my memory on this is rather hazy. (In any case, this is a finicky part of rustc. Just looking at the amount of "fix" [here](https://github.com/rust-lang/rust/blame/498553fc04f6a3fdc53412320f4e913bc53bc267/compiler/rustc_codegen_ssa/src/back/link.rs#L1315-L1322). I'd rather not touch it.)

In any case, that's not something I intend to fix here. ^^

jcaesar replied:
There are other options?

In case you're thinking of just doing
```nix
    linkerForHost = if stdenv.cc.bintools.isLLVM
      then "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}clang"
      else "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}gcc";
    # resp. doing this for all of cc/cxx/linkerForHost
```

that won't work without further adjustments either:

```
  = note: clang: error: unable to execute command: Executable "wasm-ld" doesn't exist!
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

And I really don't now what those adjustments are, or whether they're benign. I've looked into whether clang can somehow be taught a different linker command before, but couldn't find out anything.

(This would also severely increase the blast radius of this PR…)

jcaesar replied:
I dug around a bit (for options), here's a few things I found:

1. It's unlikely that rustc upstream will accept patches that recognize a `-cc` suffix as a compiler wrapping a linker, reason being that `-cc` might be anything, so it's unclear what arguments can be passed to it
2. cargo-auditable passes linker flags with `-Wl,` unconditionally ([src](https://github.com/rust-secure-code/cargo-auditable/blob/38b37332dfa3601f90608c1565fbd6e63bd6b08a/cargo-auditable/src/rustc_wrapper.rs#L64)) meaning we can't just use a plain linker as `linkerForHost`, it has to be invoked through a compiler as long as we want to use cargo-auditable.
3. 1 and 2 combined mean that some `if stdenv.cc.bintools.isLLVM then clang else gcc` is necessary for `linkerFor…`.
4. Using clang as a linker may work, but it's mechanism for finding wasm-ld is not nice: It searches paths passed with `-B` for `<triple>-wasm-ld` and `wasm-ld` ([src](https://github.com/llvm/llvm-project/blob/ad7aeb0ff58ebd29f68adb85c64e8010639e2a76/clang/lib/Driver/Driver.cpp#L6219-L6226)). However, it doesn't properly normalize the triple, and just uses the `-target wasm32-wasi` argument it gets from rustc as is. Clang itself doesn't think that's a good state and does have a [TODO](https://github.com/llvm/llvm-project/blob/ad7aeb0ff58ebd29f68adb85c64e8010639e2a76/clang/lib/Driver/Driver.cpp#L6187) about this. 
  (This might also be fixable in rustc, which does seem to have logic to not directly pass its own target name. But I havent looked into how that works.)
5. The way the linker gets passed from bintools to clang is:
   * LLVM bintools gets compiled
   * bintools-wrapper wraps it (currently not including wasm-ld) ([src](https://github.com/NixOS/nixpkgs/blob/999be9b3390bbff39f6adb9335f215c2b035744a/pkgs/build-support/bintools-wrapper/default.nix#L242-L245))
   * cc-wrapper symlinks that into ints own bin directory ([src](https://github.com/NixOS/nixpkgs/blob/f8b69469ae63cd405b22bff2a949ba5721a21017/pkgs/build-support/cc-wrapper/default.nix#L371-L374))
   * cc-wrapper's `add-flags.sh` passes the `-B` flag to its own bin dir ([src](https://github.com/NixOS/nixpkgs/blob/d1ae288c68a3a3455e46cbd7c35c0babb3485c79/pkgs/build-support/cc-wrapper/add-flags.sh#L30-L31))
6. 4+5 could in theory be worked around like [this](https://github.com/user-attachments/files/16284747/changerustlinker.patch.txt), but that's a far more terrible hack than what I've tried to do here.

So, the only way I see to get this to work cleanly is to submit a fix to clang for triple normalization, and then submit a PR to nixpkgs staging that switches to `if stdenv.cc.bintools.isLLVM then clang else gcc` (oh the rebuilds, I don't have hardware to test this), and then finally get back to this. I think that's above the level of difficulty I can manage.  

If you've got any advice on how else I could approach this, please do tell.

jcaesar committed:
wasi32: fix linking with clang invoked from rustc

Artturin commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -146,14 +146,16 @@ in stdenv.mkDerivation (finalAttrs: {
     # doesn't work) to build a linker.
     "--disable-llvm-bitcode-linker"
-  ] ++ optionals (stdenv.isLinux && !stdenv.targetPlatform.isRedox) [
+  ] ++ optionals (stdenv.isLinux && !stdenv.targetPlatform.isRedox && !stdenv.targetPlatform.isWasi) [
```
Can be dropped after https://github.com/NixOS/nixpkgs/pull/343059

Artturin replied:
It's been merged btw

jcaesar replied:
Thank you for letting me know, but I'm a bit at a loss with the changes to `rust/lib/default.nix`. (I have an idea, but it'll require mass rebuilds. :/)

jcaesar committed:
rustc,rustPlatform: adjust compiler and linker names to match and be recognizable by rust

rustc infers the linker flavor from the name of the linker, but refuses to recognize "cc" as anything.
The reasons for this are not well documented, but are hinted at in a comment:
https://github.com/rust-lang/rust/blob/18deb53874ac4701ba10ebc016cb8cde7a049b82/compiler/rustc_codegen_ssa/src/back/link.rs#L1392-L1397
In short, "cc" could be (or at least historically could have been) anything, and there's no way to know what flags to pass to it.

This is a problem because nixpkgs insists to on overriding CARG_TARGET_…_LINKER to gcc or clang always,
but rustc defaults to expecting ldd directly (not wrapped in a compiler) for some targets (e.g. wasi).
Since rustc did not detec this override changing the flaver, linking has so far been broken for these targets.

Adjusting rustc itself is not absolutely necessary, but a matter of sanity.

jcaesar committed:
rustc: wasi32: disable expecting self-contained linker, the nixpkgs derivation does not supply one

jcaesar committed:
rustc: do not use clang-built llvm on wasi, as it doesn't build and is unnecessary

jcaesar commented:
Here I'd thought I could finally make progress because found a decent workaround for the mess in the rust build environment setup (namely: do actually use linkers that are either named `*-gcc` or `*-clang` (just like actually passed to rustc). There's a longer rationale in the commit message.)

And already, three new messes have appeared.
* There's now three wasi targets (-p1, p1-threads, -p2), and that isn't being dealt with anywhere yet, at all
* rustc now uses llvm fired clang, which doesn't build for wasi. (I don't fully understand what's going on)
* rustc started to expect wasi targets to use a built-in linker (i.e. placed at `…/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld/`). Fixed by a single substituteInPlace, but :/. (Placing a linker in that path is possible, but annoying since `x86_64-unknown-linux-gnu` is a symlink to a different derivation thanks to fastCross)

I think I'll let this rest.
